### PR TITLE
8009550: PlatformPCSC should load versioned so

### PR DESCRIPTION
**Pull request by gnu-andrew**
There is a long standing limitation in the UNIX smartcardio implementation which means it will only look for the `pcsclite` library in two locations; `/usr/lib` and `/usr/local/lib`. It also only searches for an unversioned library.

On systems that separate libraries from development headers in separate subpackages, the unversioned library is only installed by the `-devel` package, which rarely happens unless the user wants to write their own application using the libpcsclite headers or build someone else's code themselves. An example of this is the [pcsc-lite-libs package on Fedora](https://koji.fedoraproject.org/koji/rpminfo?rpmID=35258843)

On Debian-based systems, there is also the issue that libraries are now placed inside a subdirectory using a [tuple](https://wiki.debian.org/Multiarch/Tuples) consisting of the instruction set, syscall ABI and libc e.g. `x86_64-linux-gnu`.  So it will not only fail to find the library when only the versioned library is installed (e.g. [libpcsclite1](https://packages.ubuntu.com/kinetic/amd64/libpcsclite1/filelist)) but also when the dev package is also installed (e.g. [libpcsclite-dev](https://packages.ubuntu.com/kinetic/amd64/libpcsclite-dev/filelist)), because the file is in e.g. `/usr/lib/x86_64-linux-gnu/libpcsclite.so`

This patch makes both cases work. It prefers versioned libraries, as potentially a `libpcsclite.so.2` could become available with a different ABI that the JDK can not work with.  All cases that worked before with just `libpcsclite.so` will still work, but it should also now pick up the library on many additional systems.

A much simpler patch (search for the versioned symlink instead of unversioned) has been [in the Fedora and RHEL packages for nearly a decade](https://bugzilla.redhat.com/show_bug.cgi?id=910107), and an earlier version of this patch with Debian support has been used in IcedTea builds since 2015. It would be nice to finally fix this upstream.

Testing:

Test case:
~~~
import java.lang.reflect.Method;

public class TestPlatformPCSC {
    public static void main(String[] args)
    throws Exception {
        Class<?> cls = Class.forName("sun.security.smartcardio.PlatformPCSC");
        Method m = cls.getDeclaredMethod("getLibraryName");
        m.setAccessible(true);
        System.err.println(m.invoke(null));
    }
}
~~~

1. With `libpcsclite.so.1` available:
~~~
$ ~/builder/trunk/images/jdk/bin/java --add-opens java.smartcardio/sun.security.smartcardio=ALL-UNNAMED -Djava.security.debug=pcsc TestPlatformPCSC
pcsc: Looking for /usr/lib64/libpcsclite.so.1
pcsc: Using PC/SC library: /usr/lib64/libpcsclite.so.1
pcsc: Looking for /usr/lib64/libpcsclite.so.1
/usr/lib64/libpcsclite.so.1
~~~

2. With just `libpcsclite.so` available:
~~~
$ ~/builder/trunk/images/jdk/bin/java --add-opens java.smartcardio/sun.security.smartcardio=ALL-UNNAMED -Djava.security.debug=pcsc TestPlatformPCSC
pcsc: Looking for /usr/lib64/libpcsclite.so.1
pcsc: Looking for /usr/lib64/libpcsclite.so.0
pcsc: Looking for /usr/lib64/libpcsclite.so
pcsc: Using PC/SC library: /usr/lib64/libpcsclite.so
~~~

4. With `libpcsclite.so.1` in the Debian-style structure:
~~~
$ ~/builder/trunk/images/jdk/bin/java --add-opens java.smartcardio/sun.security.smartcardio=ALL-UNNAMED -Djava.security.debug=pcsc TestPlatformPCSC
pcsc: Looking for /usr/lib64/libpcsclite.so.1
pcsc: Looking for /usr/lib64/libpcsclite.so.0
pcsc: Looking for /usr/lib64/libpcsclite.so
pcsc: Looking for /usr/local/lib64/libpcsclite.so.1
pcsc: Looking for /usr/local/lib64/libpcsclite.so.0
pcsc: Looking for /usr/local/lib64/libpcsclite.so
pcsc: Looking for /usr/lib/x86_64-linux-gnu/libpcsclite.so.1
pcsc: Using PC/SC library: /usr/lib/x86_64-linux-gnu/libpcsclite.so.1
~~~

5. Without the library installed
~~~
$ ~/builder/trunk/images/jdk/bin/java --add-opens java.smartcardio/sun.security.smartcardio=ALL-UNNAMED -Djava.security.debug=pcsc TestPlatformPCSC
pcsc: Looking for /usr/lib64/libpcsclite.so.1
pcsc: Looking for /usr/lib64/libpcsclite.so.0
pcsc: Looking for /usr/lib64/libpcsclite.so
pcsc: Looking for /usr/local/lib64/libpcsclite.so.1
pcsc: Looking for /usr/local/lib64/libpcsclite.so.0
pcsc: Looking for /usr/local/lib64/libpcsclite.so
pcsc: Looking for /usr/lib/x86_64-linux-gnu/libpcsclite.so.1
pcsc: Looking for /usr/lib/x86_64-linux-gnu/libpcsclite.so.0
pcsc: Looking for /usr/lib/x86_64-linux-gnu/libpcsclite.so
pcsc: Looking for /usr/lib/arm-linux-gnueabi/libpcsclite.so.1
pcsc: Looking for /usr/lib/arm-linux-gnueabi/libpcsclite.so.0
pcsc: Looking for /usr/lib/arm-linux-gnueabi/libpcsclite.so
pcsc: Looking for /usr/lib/arm-linux-gnueabihf/libpcsclite.so.1
pcsc: Looking for /usr/lib/arm-linux-gnueabihf/libpcsclite.so.0
pcsc: Looking for /usr/lib/arm-linux-gnueabihf/libpcsclite.so
pcsc: Looking for /usr/lib/x86_64-kfreebsd-gnu/libpcsclite.so.1
pcsc: Looking for /usr/lib/x86_64-kfreebsd-gnu/libpcsclite.so.0
pcsc: Looking for /usr/lib/x86_64-kfreebsd-gnu/libpcsclite.so
pcsc: Looking for /usr/lib64/libpcsclite.so.1
pcsc: Looking for /usr/lib64/libpcsclite.so.0
pcsc: Looking for /usr/lib64/libpcsclite.so
pcsc: Looking for /usr/local/lib64/libpcsclite.so.1
pcsc: Looking for /usr/local/lib64/libpcsclite.so.0
pcsc: Looking for /usr/local/lib64/libpcsclite.so
pcsc: Looking for /usr/lib/x86_64-linux-gnu/libpcsclite.so.1
pcsc: Looking for /usr/lib/x86_64-linux-gnu/libpcsclite.so.0
pcsc: Looking for /usr/lib/x86_64-linux-gnu/libpcsclite.so
pcsc: Looking for /usr/lib/arm-linux-gnueabi/libpcsclite.so.1
pcsc: Looking for /usr/lib/arm-linux-gnueabi/libpcsclite.so.0
pcsc: Looking for /usr/lib/arm-linux-gnueabi/libpcsclite.so
pcsc: Looking for /usr/lib/arm-linux-gnueabihf/libpcsclite.so.1
pcsc: Looking for /usr/lib/arm-linux-gnueabihf/libpcsclite.so.0
pcsc: Looking for /usr/lib/arm-linux-gnueabihf/libpcsclite.so
pcsc: Looking for /usr/lib/x86_64-kfreebsd-gnu/libpcsclite.so.1
pcsc: Looking for /usr/lib/x86_64-kfreebsd-gnu/libpcsclite.so.0
pcsc: Looking for /usr/lib/x86_64-kfreebsd-gnu/libpcsclite.so
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at TestPlatformPCSC.main(TestPlatformPCSC.java:9)
Caused by: java.io.IOException: No PC/SC library found on this system
~~~

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8009550](https://bugs.openjdk.org/browse/JDK-8009550): PlatformPCSC should load versioned so (**Bug** - P3)

### Reviewers
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**) ⚠️ Review applies to [d0523302](https://git.openjdk.org/jdk/pull/15409/files/d0523302416bc6507696f20d1068f16427bcf6b8)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [d0523302](https://git.openjdk.org/jdk/pull/15409/files/d0523302416bc6507696f20d1068f16427bcf6b8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15409/head:pull/15409` \
`$ git checkout pull/15409`

Update a local copy of the PR: \
`$ git checkout pull/15409` \
`$ git pull https://git.openjdk.org/jdk.git pull/15409/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15409`

View PR using the GUI difftool: \
`$ git pr show -t 15409`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15409.diff">https://git.openjdk.org/jdk/pull/15409.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15409#issuecomment-1690832191)

### CODE REVIEW

**Review by valeriepeng — COMMENTED** [review]
Changes look fine. I submitted a mach5 test job just in case. Will approve once the test job passes. 
BTW, I added a noreg-other label since there is no regression test for this change.

**Review by valeriepeng — APPROVED** [review]

**Comment by tstuefe** [comment]
Hi @gnu-andrew,

in your last example, why does it look for both arm and x64 packages? And why for kFreeBsd? I see you have both hardcoded, why?

I would expect it only to attempt and pick up the architecture and OS the VM was built for.

Cheers, Thomas

**Comment by gnu-andrew** [comment]
> Hi @gnu-andrew,
> 
> in your last example, why does it look for both arm and x64 packages? And why for kFreeBsd? I see you have both hardcoded, why?
> 
> I would expect it only to attempt and pick up the architecture and OS the VM was built for.

Good question.

Because they don't fit the template `$ARCH-linux-gnu/libpcsclite.so` which would expand to `arm-linux-gnu/libpcsclite.so`.

I don't know of a way off-hand to get the ABI or the kernel in use (`kfreebsd` is not a BSD variant, but the usual GNU userland paired with the FreeBSD kernel rather than Linux). The [Wikipedia page](https://en.wikipedia.org/wiki/Debian#Derivatives_and_flavors) actually says it's now discontinued, so maybe we can just drop that one. It probably shows how long ago I [originally wrote and tested these paths](https://icedtea.wildebeest.org/hg/release/icedtea7-forest-2.6/jdk/rev/ae5765c7b8e2)... :)

In short, that was my lazy option for catching those cases that won't fit the common one. I'm open to suggestions. We could skip any template with `'arm'` in, I guess, if the architecture doesn't match. It is worth noting though, that this file is already common to all the UNIX platforms and doesn't do any OS checks, despite the last check being a MacOS framework. That also presumably means MacOS doesn't have any of the `/usr` libraries in turn
.
> 
> Cheers, Thomas

**Comment by gnu-andrew** [comment]
> Changes look fine. I submitted a mach5 test job just in case. Will approve once the test job passes. BTW, I added a noreg-other label since there is no regression test for this change.

Thanks. Yes, I don't see how we can have a regression test for this one. My own testing was manual and involved moving the system library around as root...

**Comment by tstuefe** [comment]
> > Hi @gnu-andrew,
> > in your last example, why does it look for both arm and x64 packages? And why for kFreeBsd? I see you have both hardcoded, why?
> > I would expect it only to attempt and pick up the architecture and OS the VM was built for.
> 
> Good question.
> 
> Because they don't fit the template `$ARCH-linux-gnu/libpcsclite.so` which would expand to `arm-linux-gnu/libpcsclite.so`.
> 
> I don't know of a way off-hand to get the ABI or the kernel in use (`kfreebsd` is not a BSD variant, but the usual GNU userland paired with the FreeBSD kernel rather than Linux). The [Wikipedia page](https://en.wikipedia.org/wiki/Debian#Derivatives_and_flavors) actually says it's now discontinued, so maybe we can just drop that one. It probably shows how long ago I [originally wrote and tested these paths](https://icedtea.wildebeest.org/hg/release/icedtea7-forest-2.6/jdk/rev/ae5765c7b8e2)... :)
> 
> In short, that was my lazy option for catching those cases that won't fit the common one. I'm open to suggestions. We could skip any template with `'arm'` in, I guess, if the architecture doesn't match. It is worth noting though, that this file is already common to all the UNIX platforms and doesn't do any OS checks, despite the last check being a MacOS framework. That also presumably means MacOS doesn't have any of the `/usr` libraries in turn .

Yes, that's a bit tricky. I was concerned about the JVM picking up the wrong library on a mulitarch system, since having multiple  of these directories is the point of multiarch.

But maybe its fine. The difference between the arm variants is that the float mode (soft vs hard) and I believe the former should always work, so if we accidentally pick it up, it should be no problem.

The kfreebsd one I'd just drop.

**Review by tstuefe — APPROVED** [review]

**Comment by gnu-andrew** [comment]
> Yes, that's a bit tricky. I was concerned about the JVM picking up the wrong library on a mulitarch system, since having multiple of these directories is the point of multiarch.
> 

If the architecture is not 32-bit arm and the library is installed for that architecture, it should catch that first via the `/usr/lib/$ARCH-linux-gnu/libpcsclite.so` template and return. The only potential issue seems to be if, for some reason, an unusable arm32 pcsclite library is installed and not one for the architecture itself, which seems very unlikely.

> But maybe its fine. The difference between the arm variants is that the float mode (soft vs hard) and I believe the former should always work, so if we accidentally pick it up, it should be no problem.

Good to know. If there is a way to pick it up, I'd be happy to add it, but I'm not aware of it and don't have access to a Debian arm32 system to test on.

> 
> The kfreebsd one I'd just drop.

Done.

**Comment by gnu-andrew** [comment]
@tstuefe can I get a final ok on this one to push? Thanks.

**Comment by tstuefe** [comment]
@gnu-andrew still looks okay to me. Thanks for taking my suggestion. Sorry for overlooking this one.

**Comment by gnu-andrew** [comment]
No problem. I know you were away for a while.
/integrate
